interfaces/utils/path_patterns.go

```diff
@@ -0,0 +1,154 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package utils
+
+import (
+	"fmt"
+	"regexp"
+)
+
+type PathPattern struct {
+	pattern string
+	regex   *regexp.Regexp
+}
+
+const maxGroupDepth = 50
+
+// createRegex converts the apparmor-like glob sequence into a regex. Loosely
+// using this as reference:
+// https://gitlab.com/apparmor/apparmor/-/blob/master/parser/parser_regex.c#L107
+func createRegex(pattern string) (string, error) {
+	regex := "^"
+
+	escapeNext := false
+	currentGroupLevel := 0
+	inCharClass := false
+	skipNext := false
+	itemCountInGroup := new([maxGroupDepth + 1]int)
+	for i, ch := range pattern {
+		if escapeNext {
+			regex += regexp.QuoteMeta(string(ch))
+			escapeNext = false
+			continue
+		}
+		if skipNext {
+			skipNext = false
+			continue
+		}
+		if inCharClass && ch != '\\' && ch != ']' {
+			// no characters are special other than '\' and ']'
+			regex += regexp.QuoteMeta(string(ch))
+			continue
+		}
+		switch ch {
+		case '\\':
+			escapeNext = true
+		case '*':
+			if regex[len(regex)-1] == '/' {
+				// if the * is at the end of the pattern or is followed by a
+				// '/' we don't want it to match an empty string:
+				// /foo/* -> should not match /foo/
+				// /foo/*bar -> should match /foo/bar
+				// /*/foo -> should not match //foo
+				pos := i + 1
+				for len(pattern) > pos && pattern[pos] == '*' {
+					pos++
+				}
+				if len(pattern) <= pos || pattern[pos] == '/' {
+					regex += "[^/]"
+				}
+			}
+
+			if len(pattern) > i+1 && pattern[i+1] == '*' {
+				// Handle **
+				regex += ".*"
+				skipNext = true
+			} else {
+				regex += "[^/]*"
+			}
+		case '?':
```

Comment:
so an example of how glob_default and glob_null differ

```
glob_default:  /foo/** will convert to /foo/[^/\x00]   that is it won't match / or the null character
glob_null:       /foo/** will convert to /foo/[^/] that is it won't match / but it will match the null character
```

This alternate conversion is only used in a couple of places where where raw_data can be used. Currently there are only two places this is used, in xattr attachment matching, and in matching abstract socket names. I think for the moment it is best to just document that glob_null isn't currently supported.

Reply:
I actually realized that the current code was implementing the glob_null logic. Since the change was not difficult, I actually added support for both globbing modes, though the PathPattern class will always use glob_default; the glob_null mode is not exposed, but it's still tested.